beacon_node/beacon_chain/src/fetch_blobs.rs

```diff
@@ -0,0 +1,226 @@
+    block_root: Hash256,
+    block: Arc<SignedBeaconBlock<T::EthSpec, FullPayload<T::EthSpec>>>,
+    publish_fn: impl FnOnce(BlobsOrDataColumns<T::EthSpec>) + Send + 'static,
+) -> Result<(), BlockError<T::EthSpec>> {
+    let versioned_hashes =
+        if let Ok(kzg_commitments) = block.message().body().blob_kzg_commitments() {
+            kzg_commitments
+                .iter()
+                .map(kzg_commitment_to_versioned_hash)
+                .collect()
+        } else {
+            vec![]
+        };
+    let num_blobs = versioned_hashes.len();
+
+    if versioned_hashes.is_empty() {
+        debug!(chain.log, "Blobs from EL - none required");
+        return Ok(());
+    }
+
+    let execution_layer = chain
+        .execution_layer
+        .as_ref()
+        .ok_or(BeaconChainError::ExecutionLayerMissing)?;
+
+    debug!(
+        chain.log,
+        "Blobs from EL - start request";
+        "num_blobs" => num_blobs,
+    );
+    let response = execution_layer
+        .get_blobs(versioned_hashes)
+        .await
+        .map_err(|e| BlockError::ExecutionPayloadError(ExecutionPayloadError::RequestFailed(e)))?;
+    let num_fetched_blobs = response.iter().filter(|b| b.is_some()).count();
+    let mut all_blobs_fetched = false;
+    if num_fetched_blobs == 0 {
+        debug!(chain.log, "Blobs from EL - response with none");
+        return Ok(());
+    } else if num_fetched_blobs < num_blobs {
+        debug!(
+            chain.log,
+            "Blobs from EL - response with some";
+            "fetched" => num_fetched_blobs,
+            "total" => num_blobs,
+        );
+    } else {
+        all_blobs_fetched = true;
+        debug!(
+            chain.log,
+            "Blobs from EL - response with all";
+            "num_blobs" => num_blobs
+        );
+    }
+
+    let (signed_block_header, kzg_commitments_proof) =
+        block.signed_block_header_and_kzg_commitments_proof()?;
+
+    let mut fixed_blob_sidecar_list = FixedBlobSidecarList::default();
+    for (i, blob_and_proof) in response
+        .into_iter()
+        .enumerate()
+        .filter_map(|(i, opt_blob)| Some((i, opt_blob?)))
+    {
+        match BlobSidecar::new_with_existing_proof(
+            i,
+            blob_and_proof.blob,
+            &block,
+            signed_block_header.clone(),
+            &kzg_commitments_proof,
+            blob_and_proof.proof,
+        ) {
+            Ok(blob) => {
+                if let Some(blob_mut) = fixed_blob_sidecar_list.get_mut(i) {
+                    *blob_mut = Some(Arc::new(blob));
+                } else {
+                    error!(
+                        chain.log,
+                        "Blobs from EL - out of bounds";
+                        "i" => i
+                    );
+                }
+            }
+            Err(e) => {
+                warn!(
+                    chain.log,
+                    "Blobs from EL - error";
+                    "error" => ?e
+                );
```

Comment:
Why not error out in these two statements? Should not happen and make the code below harder to reason about

Reply:
Done